### PR TITLE
xds: increment config_reload after workers receive TLS update

### PR DESCRIPTION
**Pull request by kyessenov**
Commit Message: partial fix for #12326:
```
//test/integration:extension_discovery_integration_test                  FAILED in 2 out of 5000 in 15.1s
 ```
Risk Level: low
Testing: unit
Docs Changes:
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]

### CODE REVIEW

**Comment by kyessenov** [comment]
I'm a little confused by the worker count. Does TLS update only listener workers, or all threads (including upstream thread?).

**Comment by mattklein123** [comment]
> I'm a little confused by the worker count. Does TLS update only listener workers, or all threads (including upstream thread?).

TLS is all workers + master thread.

File: test/common/filter/http/filter_config_discovery_impl_test.cc [hunk 1]
```diff
@@ -51,6 +52,7 @@ class FilterConfigDiscoveryTestBase : public testing::Test {
     ON_CALL(init_manager_, initialize(_))
         .WillByDefault(Invoke(
             [this](const Init::Watcher& watcher) { init_target_handle_->initialize(watcher); }));
+    EXPECT_CALL(factory_context_.admin_, concurrency()).WillRepeatedly(Return(0));
```
Comment (mattklein123) [inline test/common/filter/http/filter_config_discovery_impl_test.cc:55]: nit: can you just make this a nice mock?

File: include/envoy/server/admin.h [hunk 2]
```diff
@@ -153,6 +153,11 @@ class Admin {
    * @param handler the handler that will receive this Admin's listener.
    */
   virtual void addListenerToHandler(Network::ConnectionHandler* handler) PURE;
+
+  /**
+   * @return the number of worker threads to run in the server.
+   */
+  virtual uint32_t concurrency() const PURE;
```
Comment (mattklein123) [inline include/envoy/server/admin.h:160]: It feels slightly strange to hang this off of admin, but I don't feel very strongly about it. Should this just go directly on the server/factory context somewhere? wdyt?

File: test/common/filter/http/filter_config_discovery_impl_test.cc (hunk 1 above)
Comment (kyessenov) [reply test/common/filter/http/filter_config_discovery_impl_test.cc:55]: It is a nice mock but it defaults to 1 for integration test purposes. Having value 0 is weird but it is what it is in the unit tests.

File: include/envoy/server/admin.h (hunk 2 above)
Comment (kyessenov) [reply include/envoy/server/admin.h:160]: This was just smallest delta. It looks like there are many implementations of a factory context which all need to be updated. I am not sure what the best place to hang it off.

File: test/common/filter/http/filter_config_discovery_impl_test.cc (hunk 1 above)
Comment (mattklein123) [reply test/common/filter/http/filter_config_discovery_impl_test.cc:55]: OK can you add a comment and switch this to ON_CALL/WillByDefault?

**Comment by kyessenov** [comment]
1 flake was a crash in postcb dispatcher call. 1 flake was the same as this was supposed to fix. It's very hard to reproduce the flakes, but I'll keep it running to get some better traces.

File: include/envoy/server/admin.h (hunk 2 above)
Comment (mattklein123) [reply include/envoy/server/admin.h:160]: I don't feel very strongly about it one way or the other. The main thing I'm concerned about is if we need to expose more options, but it's probably fine to put these here for now. @alyssawilk any thoughts on this?

**Comment by kyessenov** [comment]
Tsan failure is unrelated.
I have tried running with Asan to get more details, but it just passes all the time:
```
//test/integration:extension_discovery_integration_test                  PASSED in 42.3s
  Stats over 5000 runs: max = 42.3s, min = 22.4s, avg = 34.4s, dev = 1.3s
```

**Comment by kyessenov** [comment]
Also passes without Asan: 
//test/integration:extension_discovery_integration_test                  PASSED in 15.8s
  Stats over 5000 runs: max = 15.8s, min = 10.3s, avg = 14.2s, dev = 0.3s

I think this is probably good enough.

File: test/common/filter/http/filter_config_discovery_impl_test.cc [hunk 3]
```diff
@@ -51,6 +52,8 @@ class FilterConfigDiscoveryTestBase : public testing::Test {
     ON_CALL(init_manager_, initialize(_))
         .WillByDefault(Invoke(
             [this](const Init::Watcher& watcher) { init_target_handle_->initialize(watcher); }));
+    // Thread local storage assumes a single (master) thread with no workers.
```
Comment (mattklein123) [inline test/common/filter/http/filter_config_discovery_impl_test.cc:55]:
```suggestion
    // Thread local storage assumes a single (main) thread with no workers.
```

**Pull request merged**